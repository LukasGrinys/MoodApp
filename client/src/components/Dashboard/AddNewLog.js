import React, {Component} from 'react';
import FontAwesome from 'react-fontawesome'
import { connect } from 'react-redux';
import { canUserLog } from '../../actions';
import styles from './AddNewLog.module.scss';
import LoadingNetItem from '../Loading/loadingNetItem';
import { Link } from 'react-router-dom';
import { getDate, getDaytime } from '../../util/dateHelpers';
import { daytimes } from '../../constants/daytimes';
import { routerRoutes } from '../../constants/routerRoutes';
import classNames from 'classnames';

class AddNewLog extends Component {
    componentWillMount() {
        const date = getDate();
        const timing = getDaytime();
        this.props.dispatch(canUserLog(date, timing, this.props.userId))
    };

    returnTimeRemaining() {
        let currentTiming = getDaytime();
        let timeOfNextLog = 0;
        
        switch(currentTiming) {
            case daytimes.morning:
                timeOfNextLog = 13;
                break;
            case daytimes.afternoon :
                timeOfNextLog = 20;
                break;
            case daytimes.evening :
                timeOfNextLog = 7;
                break;
            default:
                timeOfNextLog = 7;
        };

        return `${timeOfNextLog}:00`
    }

    render(){
        const { canLog, darkTheme } = this.props;

        if (canLog === undefined) {
            return <LoadingNetItem/>
        } 

        if (canLog && getDaytime() !== daytimes.night) {
            return (
                <Link to={routerRoutes.addNewLog}>
                    <div className={classNames(
                        styles.logbutton, 
                        styles.available,
                        darkTheme && styles.dark
                        )}>
                        <div className={styles.attentionBorder}/>
                        <FontAwesome name="fas fa-plus" className={styles.floating}/> ADD LOG
                    </div>
                </Link> 
            )
        }

        return (
            <div className={classNames(styles.logbutton, styles.unavailable)}>
                Next log will be available at: {this.returnTimeRemaining()}
            </div>
        )
    }
};

function mapStateToProps({logs, user}) {
    const { id : userId } = user;
    const { canLog } = logs; 

    return { canLog, userId }
}

export default connect(mapStateToProps)(AddNewLog);