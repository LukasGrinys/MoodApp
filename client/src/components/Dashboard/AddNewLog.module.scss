@import '../../styles/config/variables';

.logbutton {
    max-width: 500px;
    border: 3px solid $color-alto-white;
    border-radius: 4px;
    padding: 10px;
    text-align: center;
    font-size: 1.2em;
    position: relative;
    margin-left: auto;
    margin-right: auto;

    &.available {
        background: #16774f;
        border-color: rgb(212, 212, 212);
        color: white;
        transition-duration: 1s;

        &:hover {
            cursor: pointer;
            border-color: rgb(60, 108, 212);
        }

        &.dark {
            background: $color-night-grey;
        }
    }

    &.unavailable {
        color:#d1d1d1;
        font-size: 1em;
        border-style: dashed;
    }
}

.floating {
    animation-name: floating;
    animation-duration: 3s;
    animation-iteration-count: infinite;
    margin-right: 8px;
    padding: 4px;
}

.attentionBorder {
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    border: 1.5px solid rgba(0, 132, 255, 0.582);
    border-radius: 4px;
    animation-name: at-border;
    animation-duration: 3s;
    animation-iteration-count: infinite;
}

@keyframes floating {
    0% { transform: scale(1, 1) }
    50% { transform: scale(1.2, 1.2)}
    100% { transform: scale(1, 1) }
}

@keyframes at-border {
    0% { opacity: 0 }
    75% { opacity : 1 }
    100% { transform: scale(1.05, 1.35 ); opacity: 0}
}
