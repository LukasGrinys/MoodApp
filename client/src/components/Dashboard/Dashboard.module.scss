@import '../../styles/config/variables';

.grid {
    border: 2px solid #d1d1d105;
    border-radius: 4px;
    width: 100%;
    max-width: 500px;
    height: 100%;
    display: flex;
    flex-wrap: wrap;
    justify-content: space-around;
    align-items: space-around;
    margin: 8px auto;

    .gridTop {
        font-weight: bold;
        font-size: 1.4em;
        text-align: center;
        width: 100%;
    }

    .gridItem {
        width: 45%;
        height: 300px;
        border: 2px solid $color-grid-border;
        border-radius: 4px;
        margin: 5px;
        padding: 5px;
        position: relative;

        .gridItemHeader {
            font-size: 22px;
            text-align: center;
            font-weight: bold;
        }
    }
}

.flexCenter {
    align-items: center;
    display: flex;
    justify-content: center;
    flex-direction: column;
    padding: 24px 36px;
}

.settingsIcon {
    color: $color-primary;
    font-size: 6em;
    opacity: 0.5;
    transition: 0.2s;
}