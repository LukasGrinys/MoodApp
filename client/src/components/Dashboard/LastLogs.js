import React from 'react';
import { Link } from 'react-router-dom';
import parentStyles from './Dashboard.module.scss';
import styles from './LastLogs.module.scss';
import LoadingNetItem from '../Loading/loadingNetItem';
import { useTheme } from '../../contexts/ThemeContext';
import { useSelector } from 'react-redux';
import Button from '../Button/Button';
import classNames from 'classnames';

const LastLogs = () => {
    const darkTheme = useTheme();
    const { lastLogs, isFetchingLogs } = useSelector( ({logs}) => {
        return {
            lastLogs : logs.lastLogs,
            isFetchingLogs : logs.isFetchingLogs
        }
    });

    return (
        <div className={parentStyles.gridItem}>
            <h3 className={parentStyles.gridItemHeader}>Last logs</h3>
            {
                isFetchingLogs ? (
                    <LoadingNetItem/>
                ) : ( (!lastLogs || !lastLogs.length) ? (
                        <div className={styles.greyText}>
                            There are no logs at the moment <br/><br/>
                            Write Your first one, by pressing the green button at the top
                        </div> 
                    ) : (
                        <div className={styles.logSection}> 
                            <div className={styles.logContainer}>
                                { lastLogs.slice(0,3).map( (log, index) => {
                                    return (
                                        <div key={index} className={styles.logItem}>
                                            <div className={styles.logContent}>
                                                <div className={classNames(
                                                    styles.ratingBox,
                                                    darkTheme && styles.dark
                                                )}>
                                                    {log.rating}
                                                </div>
                                                <div className={styles.rightSection}>
                                                    <div className={styles.date}>{log.date}</div>
                                                    <div className={styles.text}>{log.text}</div>
                                                </div>
                                            </div>
                                            <Link 
                                                to={`/logs/${log.id}`}
                                                className={classNames(
                                                    styles.link,
                                                    darkTheme && styles.dark
                                                )}
                                            >
                                                Read more
                                            </Link> 
                                        </div>
                                    )
                                })}
                            </div>
                            <Link to="/logs">
                                <Button>All Logs</Button>
                            </Link>
                        </div> )
                )
            }
            
        </div>
    )
};

export default LastLogs;