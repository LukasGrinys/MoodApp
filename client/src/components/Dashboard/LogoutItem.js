import React from 'react';
import { Link } from 'react-router-dom';
import { useTheme } from '../../contexts/ThemeContext';
import { routerRoutes } from '../../constants/routerRoutes';
import parentClasses from './Dashboard.module.scss';
import styles from './LogoutItem.module.scss';
import classNames from 'classnames';
import FontAwesome from 'react-fontawesome';

const LogoutItem = () => {
    const darkTheme = useTheme();

    return (
        <div className={classNames(
            parentClasses.gridItem,
            parentClasses.shrink
            )}>
            <h3 className={parentClasses.gridItemHeader}>Log out</h3>
            <div className={parentClasses.flexCenter}>
                <Link to={routerRoutes.logout}>
                    <div className={styles.logoutIconWrapper}> 
                        <FontAwesome name="sign-out" className={classNames(
                            styles.logoutIcon,
                            darkTheme && styles.dark
                        )}/>
                    </div>
                </Link>
            </div>
        </div>
    )
};

export default LogoutItem;