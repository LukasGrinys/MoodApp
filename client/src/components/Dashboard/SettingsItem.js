import React from 'react';
import { Link } from 'react-router-dom';
import parentClasses from './Dashboard.module.scss';
import styles from './SettingsItem.module.scss';
import { routerRoutes } from '../../constants/routerRoutes';
import { useTheme } from '../../contexts/ThemeContext';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faCog } from '@fortawesome/free-solid-svg-icons';

import classNames from 'classnames';

const SettingsItem = () => {
    const darkTheme = useTheme();

    return (
        <div className={classNames(
            parentClasses.gridItem,
            parentClasses.shrink
            )}>
            <h3 className={parentClasses.gridItemHeader}>Settings</h3>
            <div className={parentClasses.flexCenter}>
                <div className={styles.settingsIconWrapper}>
                    <Link to={routerRoutes.settings}>
                        {}
                        <FontAwesomeIcon icon={faCog} className={classNames(
                            styles.settingsIcon,
                            darkTheme && styles.dark
                            )}/>
                    </Link>
                </div>    
            </div>
        </div>
    );
};

export default SettingsItem;