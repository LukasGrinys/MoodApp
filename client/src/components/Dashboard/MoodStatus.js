import React, { Fragment } from 'react';
import { Link } from 'react-router-dom';
import LoadingNetItem from '../Loading/loadingNetItem';
import { useTheme } from '../../contexts/ThemeContext';
import Button from '../Button/Button';
import parentClasses from './Dashboard.module.scss';
import styles from './MoodStatus.module.scss';
import { useMoodStatus } from '../../hooks/Dashboard/useMoodStatus';
import classNames from 'classnames';

const MoodStatus = () => {
    const darkTheme = useTheme();

    const {
        rating,
        ratingText,
        isFetchingLogs
    } = useMoodStatus();

    return (
        <div className={classNames(
            parentClasses.gridItem,
            styles.statusContainer
        )}>
            {
                !isFetchingLogs ? (
                    <Fragment>
                        <h3 className={parentClasses.gridItemHeader}>Latest mood status</h3>
                        <div className={classNames(
                            styles.statusBox,
                            darkTheme && styles.dark
                        )}>{rating}</div>
                        <div className={styles.statusText}>{ratingText}</div>
                        <Link to="/stats">
                            <Button>Stats</Button> 
                        </Link>
                    </Fragment>
                ) : (
                    <LoadingNetItem/>
                )
            }
        </div>
    )
};

export default MoodStatus;