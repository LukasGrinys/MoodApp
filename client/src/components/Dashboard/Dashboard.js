import React, { useEffect } from 'react';
import AddNewLog from './AddNewLog';
import Loading from '../Loading/loading';
import { useTheme } from '../../contexts/ThemeContext';
import { useSelector, useDispatch } from 'react-redux';
import styles from './Dashboard.module.scss';
import LastLogs from './LastLogs';
import MoodStatus from './MoodStatus';
import SettingsItem from './SettingsItem';
import LogoutItem from './LogoutItem';
import { getLastLogs } from '../../actions';

const Dashboard = () => {
    const darkTheme = useTheme();
    const isSignedIn = useSelector( ({user}) => user.isAuth );
    const userId = useSelector( ({user}) => user.id );
    const dispatch = useDispatch();

    useEffect( () => {
        if (userId) {
            dispatch(getLastLogs({
                userId,
                limit: 9
            }))
        }
    }, [userId, dispatch]);

    if (isSignedIn === undefined) {
        return (
            <Loading/>
        )
    }

    if (isSignedIn) {
        return (
            <div>
                <AddNewLog darkTheme={darkTheme}/>
                <div className={styles.grid}>
                    <h1 className={styles.gridTop}>Your dashboard</h1>
                    <LastLogs/>
                    <MoodStatus/>
                    <SettingsItem/>
                    <LogoutItem/>
                </div>
            </div>
        )
    }
}

export default Dashboard;