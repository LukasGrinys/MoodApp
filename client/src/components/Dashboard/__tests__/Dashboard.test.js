import React from 'react';
import { Provider } from 'react-redux';
import { BrowserRouter as Router } from 'react-router-dom';
import Dashboard from '../Dashboard';
import Loading from '../../Loading/loading';
import { mount } from 'enzyme';
import configureStore from 'redux-mock-store';

import LastLogs from '../LastLogs';
import MoodStatus from '../MoodStatus';
import SettingsItem from '../SettingsItem';
import LogoutItem from '../LogoutItem';

const mockStore = configureStore();

describe('Dashboard: loading', () => {
    const store = mockStore({
        user: {
            isAuth : null
        },
        logs: {
            lastLogs : {
                isFetching: true,
                logs: null
            }
        }
    });

    const wrapper = mount(
        <Provider store={store}>
            <Dashboard/>
        </Provider>
    );

    test('Renders properly', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('Has loading widget', () => {
        expect(wrapper.find(Loading).length).toBe(1);
    })
});

describe('Dashboard when loaded', () => {
    const store = mockStore({
        user: {
            isAuth : true
        },
        logs: {
            canLog: true,
            lastLogs : {
                isFetching: false,
                logs: []
            }
        }
    });

    const wrapper = mount(
        <Router>
            <Provider store={store}>
                <Dashboard/>
            </Provider>
        </Router>
    );

    test('Renders properly', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('Has LastLogs item', () => {
        expect(wrapper.find(LastLogs).length).toBe(1);
    });

    test('Has MoodStatus item', () => {
        expect(wrapper.find(MoodStatus).length).toBe(1);
    });

    test('Has SettingsItem item', () => {
        expect(wrapper.find(SettingsItem).length).toBe(1);
    });

    test('Has Logout item', () => {
        expect(wrapper.find(LogoutItem).length).toBe(1);
    });
});