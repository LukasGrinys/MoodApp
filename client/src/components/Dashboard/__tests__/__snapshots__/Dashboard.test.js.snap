// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard when loaded Renders properly 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Provider
      store={
        Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        }
      }
    >
      <Dashboard>
        <div>
          <Connect(AddNewLog)>
            <AddNewLog
              canLog={true}
              dispatch={[Function]}
              userId={null}
            >
              <Link
                to="/addlog"
              >
                <LinkAnchor
                  href="/addlog"
                  navigate={[Function]}
                >
                  <a
                    href="/addlog"
                    onClick={[Function]}
                  >
                    <div
                      className="logbutton available"
                    >
                      <div
                        className="attentionBorder"
                      />
                      <FontAwesome
                        className="floating"
                        name="fas fa-plus"
                      >
                        <span
                          aria-hidden={true}
                          className="fa fa-fas fa-plus floating"
                        />
                      </FontAwesome>
                       ADD LOG
                    </div>
                  </a>
                </LinkAnchor>
              </Link>
            </AddNewLog>
          </Connect(AddNewLog)>
          <div
            className="grid"
          >
            <h1
              className="gridTop"
            >
              Your dashboard
            </h1>
            <LastLogs>
              <div
                className="gridItem"
              >
                <h3
                  className="gridItemHeader"
                >
                  Last logs
                </h3>
                <div
                  className="greyText"
                >
                  There are no logs at the moment 
                  <br />
                  <br />
                  Write Your first one, by pressing the green button at the top
                </div>
              </div>
            </LastLogs>
            <MoodStatus>
              <div
                className="gridItem statusContainer"
              >
                <h3
                  className="gridItemHeader"
                >
                  Latest mood status
                </h3>
                <div
                  className="gridItemContent"
                >
                  <div
                    className="statusBox"
                  >
                    N/A
                  </div>
                  <div
                    className="statusText"
                  />
                  <Link
                    to="/stats"
                  >
                    <LinkAnchor
                      href="/stats"
                      navigate={[Function]}
                    >
                      <a
                        href="/stats"
                        onClick={[Function]}
                      >
                        <Button>
                          <button
                            className="button btnPrimary"
                            onClick={[Function]}
                            type="button"
                          >
                            Stats
                          </button>
                        </Button>
                      </a>
                    </LinkAnchor>
                  </Link>
                </div>
              </div>
            </MoodStatus>
            <SettingsItem>
              <div
                className="gridItem shrink"
              >
                <h3
                  className="gridItemHeader"
                >
                  Settings
                </h3>
                <div
                  className="flexCenter"
                >
                  <div
                    className="settingsIconWrapper"
                  >
                    <Link
                      to="/settings"
                    >
                      <LinkAnchor
                        href="/settings"
                        navigate={[Function]}
                      >
                        <a
                          href="/settings"
                          onClick={[Function]}
                        >
                          <FontAwesome
                            className="settingsIcon"
                            name="cog"
                          >
                            <span
                              aria-hidden={true}
                              className="fa fa-cog settingsIcon"
                            />
                          </FontAwesome>
                        </a>
                      </LinkAnchor>
                    </Link>
                  </div>
                </div>
              </div>
            </SettingsItem>
            <LogoutItem>
              <div
                className="gridItem shrink"
              >
                <h3
                  className="gridItemHeader"
                >
                  Log out
                </h3>
                <div
                  className="flexCenter"
                >
                  <Link
                    to="/logout"
                  >
                    <LinkAnchor
                      href="/logout"
                      navigate={[Function]}
                    >
                      <a
                        href="/logout"
                        onClick={[Function]}
                      >
                        <div
                          className="logoutIconWrapper"
                        >
                          <FontAwesome
                            className="logoutIcon"
                            name="sign-out"
                          >
                            <span
                              aria-hidden={true}
                              className="fa fa-sign-out logoutIcon"
                            />
                          </FontAwesome>
                        </div>
                      </a>
                    </LinkAnchor>
                  </Link>
                </div>
              </div>
            </LogoutItem>
          </div>
        </div>
      </Dashboard>
    </Provider>
  </Router>
</BrowserRouter>
`;

exports[`Dashboard: loading Renders properly 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Dashboard>
    <Loading>
      <div
        className="loadingScreen"
      >
        <div>
          Loading
        </div>
        <br />
        <div
          className="barContainer"
        >
          <div
            className="bar"
          />
          <div
            className="bar delay1"
          />
          <div
            className="bar delay2"
          />
        </div>
      </div>
    </Loading>
  </Dashboard>
</Provider>
`;
